### PR TITLE
Fix race condition when committing changes to fork choice votes

### LINKED ISSUE #2562: ProtoNode: Delta to be subtracted is greater than node weight.
### Description
The error below was displayed many times while teku seemed to run normally, however the only attestation submitted before restart and after these errors was missed , upon restart I want to note that the errors did not show up again within nearly 2 hours of operation. Additionally few of the published attestations actually succeeded.
```
15:16:00.407 ERROR - PLEASE FIX OR REPORT | Unexpected exception thrown for event 'interface tech.pegasys.teku.util.time.channels.TimeTickChannel.onTick' in handler 'tech.pegasys.teku.services.beaconchain.BeaconChainController' java.util.concurrent.CompletionException: java.lang.RuntimeException: ProtoNode: Delta to be subtracted is greater than node weight. (See log file for full stack trace)
```

### Versions
* Software version: [teku/v0.12.3-dev-f8453a95/linux-x86_64/-ubuntu-openjdk64bitservervm-java-11]

* Java version: [
openjdk version "11.0.8" 2020-07-14
OpenJDK Runtime Environment (build 11.0.8+10-post-Ubuntu-0ubuntu120.04)
OpenJDK 64-Bit Server VM (build 11.0.8+10-post-Ubuntu-0ubuntu120.04, mixed mode, sharing)]

* OS Name & Version: [VERSION="20.04.1 LTS (Focal Fossa)"] running on windows subsystem for linux (wsl)
* Windows Version:[Version 10.0.18362.535]
* Kernel Version: [Linux L-0614 4.4.0-18362-Microsoft 476-Microsoft Fri Nov 01 16:53:00 PST 2019 x86_64 x86_64 x86_64 GNU/Linux]

### Additional Information
[teku.log](https://github.com/PegaSysEng/teku/files/5059895/teku.log)

### PR DESCRIPTION
**Pull request by ajsutton**
## PR Description
Fixes race conditions in fork choice because while changes to votes were made within a read/write lock, the store transaction they were changed in was committed outside of that lock and so may wind up overwriting conflicting changes.

Also fixes access to `Store.votes` to ensure it only occurs inside the store's read lock.

ie:
When we process attestations, we start a StoreTransaction then take the votes write lock, update the votes with next nextRoot and nextEpoch, release the write lock and then commit the transaction.
When we update head, we start a StoreTransaction , take the votes write lock, apply deltas and set the vote current root to next root, release the write lock and then commit the transaction.
Trouble is we might interleave those like:
Attestation - start transaction
Attestation - take write lock, update nextRoot, release write lock
Update Head - start transaction
Update Head - take write lock, apply deltas, update current root, release write lock
Update Head - commit transaction
Attestation - commit transaction, overwriting the vote changes that Update Head made.

With this change we likely have more locks than we actually need.  `ForkChoice` has been moved to use a `ReentrantLock` instead of `synchronized` so that it can use futures safely and still ensure the lock is held for the duration of the action and guaranteed to be released at the end.  But since all it's access to `ProtoArrayForkChoiceStrategy` is now controlled by that lock, we likely don't need as many locks in `ProtoArrayForkChoiceStrategy`. It may be better to make `ProtoArrayForkChoiceStrategy` not thread safe and control access entirely through `ForkChoice` but will leave that for further investigation later once we confirm this fixes the immediate issue.

## Fixed Issue(s)
fixes #2562

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

### CODE REVIEW

**Review by rolfyone — APPROVED** [review]
LGTM, but i wouldn't be against someone else taking a look too :)

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java [hunk 1]
```diff
@@ -48,97 +52,119 @@ private void initializeProtoArrayForkChoice() {
...
-            });
+        .thenCompose(
+            targetState ->
+                withLock(
+                    () -> {
+                      final StoreTransaction transaction = recentChainData.startStoreTransaction();
+                      final AttestationProcessingResult result =
+                          on_attestation(
+                              transaction, attestation, targetState, getForkChoiceStrategy());
+                      return transaction.commit().thenApply(__ -> result);
```
Comment (cemozerr) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java]: Seems like we don't check if the attestation processing was successful before committing anymore. Is this on purpose?

**Review by cemozerr — APPROVED** [review]
LGTM. Good catch.

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java (hunk 1 above)
Comment (ajsutton) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java]: No. Good spot we need to fix that.

**Pull request merged**